internal/venice-common/src/main/java/com/linkedin/venice/utils/SharedObjectFactory.java

```diff
@@ -0,0 +1,45 @@
+package com.linkedin.venice.utils;
+
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+
+public class SharedObjectFactory<T> {
+  private final Map<String, ReferenceCounted<T>> objectMap = new VeniceConcurrentHashMap<>();
+
+  public T getObject(String objectIdentifier, Supplier<T> objectSupplier, Consumer<T> destroyer) {
+    if (!objectMap.containsKey(objectIdentifier)) {
+      T object = objectSupplier.get();
+      objectMap.put(objectIdentifier, new ReferenceCounted<>(object, obj -> {
+        destroyer.accept(obj);
+        objectMap.remove(objectIdentifier);
+      }));
+      return object;
+    } else {
+      ReferenceCounted<T> referenceCounted = objectMap.get(objectIdentifier);
+      referenceCounted.retain();
+      return referenceCounted.get();
+    }
+  }
+
+  public void release(String objectIdentifier) {
```

Comment:
Do you think we can make this function to perform the count check such that client doesn't have to maintain the `isShared` flag and can also benefit other future classes using this factory?

What I have in my mind is like:
``` Java
public boolean release(String objectIdentifier) {
    ReferenceCounted<T> referenceCounted = objectMap.get(objectIdentifier);
    if (referenceCounted != null && referenceCounted.getCount() == 0) {
      referenceCounted.release();
      return true;
    } else {
      return false; // or maybe throw exception?
    }
}
```

then in the `close` method of `ControllerClient`, we can catch this explicitly
```
    if (ControllerClientFactory.release(this)) {
      // released
    } else {
      // no released
     }
```

Reply:
Makes sense. I kind of like specifying the `destroyer` in the constructor as it reduces the state we need to carry explicitly. I guess I can remove the `isShared` flag in a different way too. Will update it

Reply:
Updated. Thanks for the suggestion. This is much cleaner